### PR TITLE
Automate email chasing referees for a response

### PR DESCRIPTION
**Pull request by davidgisbey**
## Context

We need to automate the chase email sent to candidates and referees once 5 days have passed.

This PR implements sending the email to the referee.

## Changes proposed in this pull request

- Adds a service that returns the references to be contacted
- Adds a service that sends the email to the referee and creates a ChaserSent object associated with the reference

## Guidance to review

A follow-up PR sending an email to the candidate will follow this

**Update**

This has now been refactored to use the ChaserSent model as discussed at our dev session. 

## Link to Trello card

https://trello.com/c/WFu3BfcB/843-automatically-send-chase-email-to-candidate-and-referee-after-5-days

## Things to check

- [x] This code doesn't rely on migrations in the same Pull Request
- [x] API release notes have been updated if necessary
- [x] New environment variables have been [added to the Azure config](https://github.com/DFE-Digital/apply-for-postgraduate-teacher-training#azure-hosting-devops-pipeline)

### CODE REVIEW

File: app/services/send_chase_email.rb [hunk 1]
```diff
@@ -0,0 +1,9 @@
+class SendChaseEmail
+  def perform(reference:)
+    RefereeMailer.reference_request_chaser_email(reference.application_form, reference)
+
+    reference.application_form.application_choices.each do |application_choice|
+      application_choice.update!(status: 'awaiting_references_and_chased')
```
Comment (tvararu) [inline app/services/send_chase_email.rb]:
```suggestion
      application_choice.send_chase_email!
```

This is a method that is auto-generated by the `event :send_chase_email` declaration in the state machine, it's better to use it instead of `update!`

File: app/workers/send_chase_email_to_referees_worker.rb [hunk 2]
```diff
@@ -0,0 +1,13 @@
+class SendChaseEmailToRefereesWorker
+  include Sidekiq::Worker
+
+  def perform
+    GetRefereesToChase.new.perform.each do |reference|
+      begin
+        SendChaseEmail.new.perform(refence: reference)
+      rescue StandardError => e
+        Rails.logger.warn "[DBD] ignoring reference #{reference.id}: #{e.message}"
+      end
```
Comment (tvararu) [inline app/workers/send_chase_email_to_referees_worker.rb]: Hmm, why is this needed? 🤔 

Why are we prefixing this with `[DBD]`, looks like a copy+paste typo?

Could we rescue on something more specific than `StandardError`?

File: app/services/application_state_change.rb [hunk 3]
```diff
@@ -20,6 +20,11 @@ def initialize(application_choice)
 
     state :awaiting_references do
       event :references_complete, transitions_to: :application_complete
+      event :send_chase_email, transitions_to: :awaiting_references_and_chased
+    end
+
+    state :awaiting_references_and_chased do
```
Comment (tvararu) [inline app/services/application_state_change.rb]: I think the naming of this is passable, what do others think? cc maester @duncanjbrown

Comment (duncanjbrown) [reply app/services/application_state_change.rb]: I think the name is good... but I'm not convinced this is a distinct state on the choice.

This seems like a state of the reference request, not the application choice. And we're opening the door on a world in which `:awaiting_references` is not an honest state because it will miss out some choices which are IRL awaiting references but aren't `:awaiting_references`

File: app/workers/send_chase_email_to_referees_worker.rb [hunk 4]
```diff
@@ -0,0 +1,13 @@
+class SendChaseEmailToRefereesWorker
+  include Sidekiq::Worker
+
+  def perform
+    GetRefereesToChase.new.perform.each do |reference|
+      begin
+        SendChaseEmail.new.perform(refence: reference)
```
Comment (davidgisbey) [inline app/workers/send_chase_email_to_referees_worker.rb]: Just realised this needs to be more explicit as there will be multiple chase emails
Perhaps SendRefereeChaseEmail?
Or SendFirstChaseEmailToReferee

File: app/services/get_referees_to_chase.rb [hunk 5]
```diff
@@ -0,0 +1,13 @@
+class GetRefereesToChase
```
Comment (davidgisbey) [inline app/services/get_referees_to_chase.rb]: This probably needs to be more explicit too.

File: app/services/send_chase_email.rb [hunk 6]
```diff
@@ -0,0 +1,6 @@
+class SendChaseEmail
+  def perform(reference:)
+    RefereeMailer.reference_request_chaser_email(reference.application_form, reference)
+    ChaserSent.create(chased: reference, chaser_type: :referee_mailer_reference_request_chaser_email)
```
Comment (tijmenb) [inline app/services/send_chase_email.rb]:
```suggestion
    ChaserSent.create!(chased: reference, chaser_type: :referee_mailer_reference_request_chaser_email)
```

On second look perhaps `referee_mailer_reference_request_chaser_email` is a little wordy? Perhaps just `reference_request`, since it's clear it's an email and chaser.

File: app/workers/send_chase_email_to_referees_worker.rb [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+      rescue StandardError => e
```
Comment (tijmenb) [inline app/workers/send_chase_email_to_referees_worker.rb]: What's the reason for rescuing errors here?

File: app/services/get_referees_to_chase.rb [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+  def perform
+    references = ApplicationReference.includes(:application_form).feedback_requested
+
+    references_with_applications_submitted_in_the_last_5_days = references.select do |reference|
+      reference.application_form.submitted_at < Time.zone.now - 5.days
+    end
+
+    references_with_applications_submitted_in_the_last_5_days.select do |reference|
+      reference.chasers_sent.referee_mailer_reference_request_chaser_email.blank?
+    end
```
Comment (tijmenb) [inline app/services/get_referees_to_chase.rb]: I think we could reduce this down to a single query. We could start by using a `where` for the `submitted_at` filtering. Let me know if you want a hand.

File: app/services/get_referees_to_chase.rb (hunk 5 above)
Comment (tijmenb) [reply app/services/get_referees_to_chase.rb]: I think this is fine 👍

File: spec/services/get_referees_to_chase_spec.rb [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+      application_form = create(:application_form, submitted_at: 6.days.ago)
+      reference = create(:reference, feedback_status: 'feedback_requested', application_form: application_form)
+      create(:application_choice, application_form: application_form, status: 'awaiting_references')
+
+      service = described_class.new.perform
+
+      expect(service).to eq [reference]
+    end
+
+    it 'only return sapplciation choices which are awaiting references' do
```
Comment (tijmenb) [inline spec/services/get_referees_to_chase_spec.rb]:
```suggestion
    it 'only returns application choices which are awaiting references' do
```

File: spec/workers/send_chase_email_to_referees_worker_spec.rb [hunk 10]
```diff
@@ -0,0 +1,21 @@
+require 'rails_helper'
+
+RSpec.describe SendChaseEmailToRefereesWorker do
```
Comment (tijmenb) [inline spec/workers/send_chase_email_to_referees_worker_spec.rb]: Can we create a system spec instead to test this?

File: app/services/get_referees_to_chase.rb (hunk 5 above)
Comment (davidgisbey) [reply app/services/get_referees_to_chase.rb]: ok thanks

File: app/workers/send_chase_email_to_referees_worker.rb (hunk 7 above)
Comment (davidgisbey) [reply app/workers/send_chase_email_to_referees_worker.rb]: I'd love to have a better answer than it was on the other workers. I assumed it would help debug if a job failed for whatever reason. I'll remove it

File: spec/workers/send_chase_email_to_referees_worker_spec.rb (hunk 10 above)
Comment (davidgisbey) [reply spec/workers/send_chase_email_to_referees_worker_spec.rb]: Sure, but roughly how would that look? The system specs seem to largely be user journey focussed. I'm not 100% sure how to test a job that way.

Comment (tijmenb) [reply spec/workers/send_chase_email_to_referees_worker_spec.rb]: Roughly:

- Given a candidate submits an application (create an application with references)
- When there's no response from the referees within 5 days (use timecop to call `SendChaseEmailToRefereesWorker` 5 days later) 
- The referee gets an email with a reminder (check email contents)

I think something similar to the [reject by default test](https://github.com/DFE-Digital/apply-for-postgraduate-teacher-training/blob/master/spec/system/provider_interface/provider_receives_email_when_a_new_application_has_been_submitted_spec.rb#L7-L13) would be nice. Perhaps you can [use some stuff from the referee test as well](https://github.com/DFE-Digital/apply-for-postgraduate-teacher-training/blob/master/spec/system/referee_interface/referee_refuses_to_give_a_reference_spec.rb#L9-L12).

Comment (davidgisbey) [reply spec/workers/send_chase_email_to_referees_worker_spec.rb]: Have done 👍

File: app/services/get_referees_to_chase.rb [hunk 11]
```diff
@@ -0,0 +1,9 @@
+class GetRefereesToChase
+  def perform
+    ApplicationReference
+    .joins(:application_form)
+    .feedback_requested
+    .where(['application_forms.submitted_at < ?', Time.zone.now - 5.days])
+    .select { |reference| reference.chasers_sent.reference_request.blank? }
```
Comment (tijmenb) [inline app/services/get_referees_to_chase.rb]:
```suggestion
    .where.not(id: ChaserSent.reference_request.select(:chased_id))
```

This will generate one query for this call:

```
[3] pry(main)> puts GetRefereesToChase.new.perform.to_sql
SELECT "references".* FROM "references" INNER JOIN "application_forms" ON "application_forms"."id" = "references"."application_form_id" WHERE "references"."feedback_status" = 'feedback_requested' AND (application_forms.submitted_at < '2020-01-30 15:07:56.592003') AND "references"."id" NOT IN (SELECT "chasers_sent"."chased_id" FROM "chasers_sent" WHERE "chasers_sent"."chaser_type" = 'reference_request')
```

File: app/services/send_chase_email_to_referee_and_candidate.rb [hunk 12]
```diff
@@ -1,7 +1,10 @@
 class SendChaseEmailToRefereeAndCandidate
   def self.call(application_form:, reference:)
     RefereeMailer.reference_request_chaser_email(application_form, reference).deliver
+    ChaserSent.create!(chased: reference, chaser_type: :reference_request)
+
     CandidateMailer.reference_chaser_email(application_form, reference).deliver
+    ChaserSent.create!(chased: application_form.candidate, chaser_type: :reference_request)
```
Comment (tijmenb) [inline app/services/send_chase_email_to_referee_and_candidate.rb]: I don't think we need this one, we're not using it right?

File: spec/services/get_referees_to_chase_spec.rb [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+      application_form = create(:application_form, submitted_at: 6.days.ago)
+      reference = create(:reference, feedback_status: 'feedback_requested', application_form: application_form)
+      create(:application_choice, application_form: application_form, status: 'awaiting_references')
+
+      service = described_class.new.perform
+
+      expect(service).to eq [reference]
+    end
+
+    it 'only returns applciation choices which are awaiting references' do
```
Comment (tijmenb) [inline spec/services/get_referees_to_chase_spec.rb]:
```suggestion
    it 'only returns application choices which are awaiting references' do
```

File: spec/system/referee_interface/referee_does_not_respond_spec.rb [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+RSpec.feature 'Referee does not respond within 5 days', sidekiq: true do
+  include CandidateHelper
+
+  scenario 'A chase email is sent if a referee does not respond within 5 days' do
+    FeatureFlag.activate('training_with_a_disability')
+
+    given_a_candidate_completed_an_application
+    when_the_candidate_submits_the_application
+    and_i_do_not_respond_within_5_days
+    then_i_am_sent_a_chase_email
```
Comment (tijmenb) [inline spec/system/referee_interface/referee_does_not_respond_spec.rb]: Perhaps `and_the_referee_does_not_respond_within_5_days` (we haven't explained that we're a referee)

File: app/services/send_chase_email_to_referee_and_candidate.rb (hunk 12 above)
Comment (davidgisbey) [reply app/services/send_chase_email_to_referee_and_candidate.rb]: Yea you're right 👍

**Comment by davidgisbey** [comment]
Thanks for all the help/feedback!

File: spec/services/get_referees_to_chase_spec.rb [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+RSpec.describe GetRefereesToChase do
+  describe '#perform' do
+    it 'returns referees that were sent their reference email more than 5 days ago and have not already been chased' do
+      application_form = create(:application_form, submitted_at: 6.days.ago)
+      reference = create(:reference, feedback_status: 'feedback_requested', application_form: application_form)
+      create(:application_choice, application_form: application_form, status: 'awaiting_references')
+
+      service = described_class.call
+
+      expect(service).to eq [reference]
```
Comment (C-gyorfi) [inline spec/services/get_referees_to_chase_spec.rb]:
```suggestion
      expect(described_class.call).to include reference
```

File: spec/services/get_referees_to_chase_spec.rb [hunk 16]
```diff
@@ -0,0 +1,49 @@
+require 'rails_helper'
+
+RSpec.describe GetRefereesToChase do
+  describe '#perform' do
+    it 'returns referees that were sent their reference email more than 5 days ago and have not already been chased' do
+      application_form = create(:application_form, submitted_at: 6.days.ago)
+      reference = create(:reference, feedback_status: 'feedback_requested', application_form: application_form)
+      create(:application_choice, application_form: application_form, status: 'awaiting_references')
+
+      service = described_class.call
```
Comment (C-gyorfi) [inline spec/services/get_referees_to_chase_spec.rb]:
```suggestion
```
You could just call these in the assertion as suggested below

File: spec/factories.rb [hunk 17]
```diff
@@ -3,7 +3,7 @@
     candidate
     chased_id { candidate.id }
     chased_type { candidate.class }
-    chased_type { :referee_mailer_reference_request_chaser_email }
+    chased_type { :reference_request }
```
Comment (C-gyorfi) [inline spec/factories.rb]: This is a bit unclear to mee, you defined chased_type twice, was this a mistake?

File: app/services/get_referees_to_chase.rb [hunk 18]
```diff
@@ -0,0 +1,9 @@
+class GetRefereesToChase
+  def self.call
+    ApplicationReference
+    .joins(:application_form)
+    .feedback_requested
+    .where(['application_forms.submitted_at < ?', Time.zone.now - 5.days])
```
Comment (tijmenb) [inline app/services/get_referees_to_chase.rb]: This needs to be `reference.created_at` to avoid fun after https://github.com/DFE-Digital/apply-for-postgraduate-teacher-training/pull/1279.

File: spec/system/referee_interface/referee_does_not_respond_spec.rb [hunk 19]
```diff
@@ -0,0 +1,42 @@
...
+    and_an_email_is_sent_to_the_candidate
+  end
+
+  def given_a_candidate_completed_an_application
+    candidate_completes_application_form
+  end
+
+  def when_the_candidate_submits_the_application
+    candidate_submits_application
+    @application.application_references.first.update(feedback_status: :feedback_refused)
```
Comment (tijmenb) [inline spec/system/referee_interface/referee_does_not_respond_spec.rb]:
```suggestion
    @application.application_references.first.update!(feedback_status: :feedback_refused)
```

File: spec/system/referee_interface/referee_does_not_respond_spec.rb [hunk 20]
```diff
@@ -0,0 +1,42 @@
...
+RSpec.feature 'Referee does not respond within 5 days', sidekiq: true do
+  include CandidateHelper
+
+  scenario 'A chase email is sent if a referee does not respond within 5 days' do
+    FeatureFlag.activate('training_with_a_disability')
+
+    given_a_candidate_completed_an_application
+    when_the_candidate_submits_the_application
+    and_the_referee_does_not_respond_within_5_days
+    then_i_am_sent_a_chase_email
```
Comment (tijmenb) [inline spec/system/referee_interface/referee_does_not_respond_spec.rb]:
```suggestion
    then_the_referee_am_sent_a_chase_email
```

Comment (C-gyorfi) [reply spec/system/referee_interface/referee_does_not_respond_spec.rb]:
```suggestion
    then_the_referee_sent_a_chase_email
```

File: spec/factories.rb (hunk 17 above)
Comment (davidgisbey) [reply spec/factories.rb]: Yea it was meant to be chaser type. Good spot 👍
